### ISSUE TITLE
Crashes when pipewire is down

### ISSUE THREAD

**Issue by travankor** [wwmm/easyeffects#1348]
### EasyEffects Version

6.2.1

### What package are you using?

Other (specify below)

### Distribution

Void Linux

### Describe the bug

Easyeffects crashes in two cases for me. One is when the pipewire user service is not running. The second is when I toggle the dark mode for easyeffects.

### Expected Behavior

Easyeffects does not crash. Instead of crashing, easyeffects should gracefully print an error message and exit.

### Debug Log

<details><summary>Debug Log</summary><br/>

```
#0  0x00007fed7855cc17 in g_log_structured_array () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#1  0x00007fed7855d025 in g_log_default_handler () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#2  0x00007fed7855d279 in g_logv () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#3  0x00007fed7855d50f in g_log () from /usr/lib/libglib-2.0.so.0
No symbol table info available.
#4  0x000055b499581473 in util::error(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) [clone .isra.0] (s=..., s=...) at ../src/util.cpp:29
No locals.
#5  0x000055b499537a52 in PipeManager::PipeManager (this=<optimized out>, this=<optimized out>)
    at ../src/pipe_manager.cpp:1357
        props_context = 0x55b49a6d5760
        props_sink = <optimized out>
        props_source = <optimized out>
        props_context = <optimized out>
        props_sink = <optimized out>
        props_source = <optimized out>
        _res = <optimized out>
        _f = <optimized out>
        _res = <optimized out>
        _f = <optimized out>
        ts = <optimized out>
        node = <optimized out>
        __for_range = <optimized out>
        __for_begin = <optimized out>
        __for_end = <optimized out>
#6  0x000055b499437d1f in app::on_startup (gapp=0x55b49a553120) at ../src/application.cpp:64
        self = 0x55b49a553120
#7  0x00007fed778dd919 in ?? () from /usr/lib/libgobject-2.0.so.0
No symbol table info available.
#8  0x00007fed778f64df in g_signal_emit_valist () from /usr/lib/libgobject-2.0.so.0
No symbol table info available.
#9  0x00007fed778f66bf in g_signal_emit () from /usr/lib/libgobject-2.0.so.0
No symbol table info available.
#10 0x00007fed77a083d2 in g_application_register () from /usr/lib/libgio-2.0.so.0
No symbol table info available.
#11 0x00007fed77a087ba in ?? () from /usr/lib/libgio-2.0.so.0
No symbol table info available.
#12 0x00007fed77a08afe in g_application_run () from /usr/lib/libgio-2.0.so.0
No symbol table info available.
#13 0x000055b49941b482 in main (argc=1, argv=0x7ffdedbb5e58) at ../src/easyeffects.cpp:56
        bindtext_output = <optimized out>
        app = 0x55b49a553120
        status = <optimized out>
```
</details>

<details><summary>Debug Log 2</summary><br/>

```
#0  0x00007f3bc55ecb67 in ?? () from /usr/lib/libgtk-4.so.1
#1  0x00007f3bc55ece29 in ?? () from /usr/lib/libgtk-4.so.1
#2  0x00007f3bc55ecef4 in ?? () from /usr/lib/libgtk-4.so.1
#3  0x00007f3bc59c20d4 in ?? () from /usr/lib/libgtk-4.so.1
#4  0x00007f3bc58b3397 in ?? () from /usr/lib/libgtk-4.so.1
#5  0x00007f3bc55dc13b in ?? () from /usr/lib/libgtk-4.so.1
#6  0x00007f3bc580e9ba in ?? () from /usr/lib/libgtk-4.so.1
#7  0x00007f3bc5817e04 in ?? () from /usr/lib/libgtk-4.so.1
#8  0x00007f3bc56d5d4a in ?? () from /usr/lib/libgtk-4.so.1
#9  0x00007f3bc56d2650 in ?? () from /usr/lib/libgtk-4.so.1
#10 0x00007f3bc56d304b in ?? () from /usr/lib/libgtk-4.so.1
#11 0x00007f3bc55cc827 in ?? () from /usr/lib/libgtk-4.so.1
#12 0x00007f3bc516c919 in ?? () from /usr/lib/libgobject-2.0.so.0
#13 0x00007f3bc51854df in g_signal_emit_valist () from /usr/lib/libgobject-2.0.so.0
#14 0x00007f3bc51856bf in g_signal_emit () from /usr/lib/libgobject-2.0.so.0
#15 0x00007f3bc5695c36 in ?? () from /usr/lib/libgtk-4.so.1
#16 0x00007f3bc516f94b in g_cclosure_marshal_VOID__BOXEDv () from /usr/lib/libgobject-2.0.so.0
#17 0x00007f3bc516c919 in ?? () from /usr/lib/libgobject-2.0.so.0
#18 0x00007f3bc51854df in g_signal_emit_valist () from /usr/lib/libgobject-2.0.so.0
#19 0x00007f3bc51856bf in g_signal_emit () from /usr/lib/libgobject-2.0.so.0
#20 0x00007f3bc5694527 in ?? () from /usr/lib/libgtk-4.so.1
#21 0x00007f3bc56952eb in ?? () from /usr/lib/libgtk-4.so.1
#22 0x00007f3bc569844f in ?? () from /usr/lib/libgtk-4.so.1
#23 0x00007f3bc5807e43 in ?? () from /usr/lib/libgtk-4.so.1
#24 0x00007f3bc5808179 in ?? () from /usr/lib/libgtk-4.so.1
#25 0x00007f3bc59d2b56 in ?? () from /usr/lib/libgtk-4.so.1
#26 0x00007f3bc56e3713 in ?? () from /usr/lib/libgtk-4.so.1
#27 0x00007f3bc581e5f8 in ?? () from /usr/lib/libgtk-4.so.1
#28 0x00007f3bc58ecc59 in ?? () from /usr/lib/libgtk-4.so.1
#29 0x00007f3bc591b72d in ?? () from /usr/lib/libgtk-4.so.1
#30 0x00007f3bc516c919 in ?? () from /usr/lib/libgobject-2.0.so.0
#31 0x00007f3bc5184824 in g_signal_emit_valist () from /usr/lib/libgobject-2.0.so.0
#32 0x00007f3bc51856bf in g_signal_emit () from /usr/lib/libgobject-2.0.so.0
#33 0x00007f3bc59cd686 in ?? () from /usr/lib/libgtk-4.so.1
#34 0x00007f3bc593163a in ?? () from /usr/lib/libgtk-4.so.1
#35 0x00007f3bc5de4c0b in g_main_context_dispatch () from /usr/lib/libglib-2.0.so.0
#36 0x00007f3bc5de4eb8 in ?? () from /usr/lib/libglib-2.0.so.0
#37 0x00007f3bc5de4f6f in g_main_context_iteration () from /usr/lib/libglib-2.0.so.0
#38 0x00007f3bc5297b65 in g_application_run () from /usr/lib/libgio-2.0.so.0
#39 0x0000560bffe7c482 in main (argc=1, argv=0x7ffd083efde8) at ../src/easyeffects.cpp:56
```
</details>

**Comment by wwmm** [comment]
>One is when the pipewire user service is not running.

I never tried to kill PipeWire to see what happens. But based on the gdb output `../src/pipe_manager.cpp:1357` the call to `util::error(log_tag + "could not create PipeWire context");` should have printed a warning before forcing the applicattion to exit.

>The second is when I toggle the dark mode for easyeffects.

I can not reproduce that on my computer (Arch Linux with GNOME). Considering that dark mode is entirely handled by gtk that sounds like a gtk4 bug.

**Comment by travankor** [comment]
>I never tried to kill PipeWire to see what happens. But based on the gdb output ../src/pipe_manager.cpp:1357 the call to util::error(log_tag + "could not create PipeWire context"); should have printed a warning before forcing the applicattion to exit.

I'm on an older version of pipewire where it said `pipe_manager: context connection failed`. It should still be possible to solve the crash.

>I can not reproduce that on my computer (Arch Linux with GNOME). Considering that dark mode is entirely handled by gtk that sounds like a gtk4 bug.

I will try and report this upstream to gtk then.

**Comment by wwmm** [comment]
>It should still be possible to solve the crash.

What `util::error` does is calling the glib function `g_error`. It forces the program to exit after printing the error message. It is not that it is crashing. It is being force to exit because nothing will work if we can not create a PipeWire context. That forced exit is intentional.

**Comment by wwmm** [comment]
Although some kid of beautiful window with an error message could be shown to the user I do not think it is worth the additional complexity this would add to the code. Maybe if we supported more than one audio server. But that is not going to happen. So PipeWire has to be running if EasyEffects is going to be used.

**Comment by travankor** [comment]
> What util::error does is calling the glib function g_error. It forces the program to exit after printing the error message. It is not that it is crashing. It is being forced to exit because nothing will work if we can not create a PipeWire context. It is intentional.

The process terminates with `SIGTRAP`. Is this signal needed in this case?

>So PipeWire has to be running if EasyEffects is going to be used.

Sure but pipewire can crash or be restarted, which would trigger this. Even with the service manager restarting pipewire, easyeffects can still crash.

**Comment by wwmm** [comment]
>The process terminates with SIGTRAP. Is this signal needed in this case?

Maybe not. When I started to use `g_error` my intention was only making the application finishing in case a critical error happened. As long as the application aborts it doesn't really matter the signal being used.

>Sure but pipewire can crash or be restarted, which would trigger this. Even with the service manager restarting pipewire, easyeffects can still crash.

That is true. And we also had this problem when we used Pulseaudio. What makes the restart hard to implement is that we have to load many things in a very specific order. And most of this initialization is happening once when EasyEffects starts. In order to automatically restart in case something happens to PipeWire some parts of our code will become much more complex. It is different from the situation of a simple audio player where you just recreate the audio stream object and move on. For example if PipeWire dies we have to reload our virtual devices and reinitialize all the structures that could make use of the information about these devices. What includes the code that links filters to devices.

In short it is what we should be doing. I agree. But in order to achieve that we will have to literally rewrite a lot of code and do a lot of testing to make sure new problems were not introduced. As the user can just restart the application and we have more concerning problems to fix it may take a really long time until some work is done in this direction.

**Comment by travankor** [comment]
Thank you for the detailed explanation. It seems not possible to solve this in the near future.

>Maybe not. When I started to use g_error my intention was only making the application finishing in case a critical error happened. As long as the application aborts it doesn't really matter the signal being used.

`g_error` has changed their behaviour from sending `SIGABRT` to `SIGTRAP`. https://gitlab.gnome.org/GNOME/glib/-/issues/1448

I don't know if glib provides a way to send `SIGABRT` or if just calling `abort()` is desireable after printing the error.
